src/Init/Data/String/Extra.lean

```diff
@@ -17,13 +17,69 @@ def toNat! (s : String) : Nat :=
   else
     panic! "Nat expected"
 
+def utf8DecodeChar? (a : ByteArray) (i : Nat) : Option Char := do
+  let c ← a[i]?
+  if c &&& 0x80 == 0 then
+    some ⟨c.toUInt32, .inl (Nat.lt_trans c.1.2 (by decide))⟩
+  else if c &&& 0xe0 == 0xc0 then
+    let c1 ← a[i+1]?
+    guard (c1 &&& 0xc0 == 0x80)
+    let r := ((c &&& 0x1f).toUInt32 <<< 6) ||| (c1 &&& 0x3f).toUInt32
+    guard (0x80 ≤ r)
+    -- FIXME: this check is not needed
+    if h : r < 0xd800 then some ⟨r, .inl h⟩ else none
```

Comment:
Why is it not? Why is it still there then?

Reply:
it's possible to prove from the previous line that `r` is at most 2^11, which is less than 0xd800. But proving those kind of facts about bitwise ops is not easy, especially in this very early file.

Reply:
Ah, now I get it. How about

```suggestion
    -- We could prove h from the definition of r, but it is too tedious,
    -- so for now we just check it again.
    if h : r < 0xd800 then some ⟨r, .inl h⟩ else none
```

Reply:
```suggestion
    -- TODO: Prove h from the definition of r once we have the necessary lemmas
    if h : r < 0xd800 then some ⟨r, .inl h⟩ else none
```